File: 133-clone-graph/clone-graph.js
/**
 * // Definition for a Node.
 * function _Node(val, neighbors) {
 *    this.val = val === undefined ? 0 : val;
 *    this.neighbors = neighbors === undefined ? [] : neighbors;
 * };
 */

/**
 * @param {_Node} node
 * @return {_Node}
 */
var cloneGraph = function(node) {
    if (!node) return null;

    const visited = new Map();
    
    const cloneGraphHelper = (node) => {
        if (visited.has(node)) return visited.get(node);

        const newNode = new _Node(node.val);
        visited.set(node, newNode);

        newNode.neighbors = node.neighbors.map(cloneGraphHelper);
        return newNode;
    };

    return cloneGraphHelper(node);
};

File: 133-clone-graph/clone-graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    Node* cloneGraph(Node* node) {
        vector<Node*> values(101, NULL);
        return helper(node, values);
    }

    Node* helper(Node* node,  vector<Node*>& values) {
        if(node == NULL) return NULL;
        if(values[node->val] != NULL) return values[node->val];

        Node* newNode = new Node();
        newNode->val = node->val;
        values[newNode->val] = newNode;

        for(Node* nbr: node->neighbors) {
            Node* nbrNode = helper(nbr, values);
            newNode->neighbors.push_back(nbrNode);
        } 

        return newNode;
    }
};